api/graphql/reservations/reservation_serializers.py

```diff
@@ -1167,3 +1167,137 @@ class Meta:
+                ],
+            )
+
+        return instance
+
+    def validate(self, data):
+        data = super().validate(data)
+
+        if self.instance.state != STATE_CHOICES.CONFIRMED:
+            raise ValidationErrorWithCode(
+                "Only reservations in confirmed state can be cancelled through this.",
+                ValidationErrorCodes.RESERVATION_MODIFICATION_NOT_ALLOWED,
+            )
+
+        if self.instance.handled_at:
+            raise ValidationErrorWithCode(
+                "Reservation has gone through handling and it cannot be changed anymore.",
+                ValidationErrorCodes.RESERVATION_MODIFICATION_NOT_ALLOWED,
+            )
+        begin = data["begin"]
+        end = data["end"]
+        self.check_begin(begin, end)
+
+        for reservation_unit in self.instance.reservation_unit.all():
+            self.check_cancellation_rules(reservation_unit)
+            self.check_reservation_time(reservation_unit, begin, end)
+            self.check_reservation_overlap(reservation_unit, begin, end)
+            self.check_reservation_duration(reservation_unit, begin, end)
+            self.check_buffer_times(reservation_unit, begin, end)
+            self.check_reservation_days_before(begin, reservation_unit)
+
+            scheduler = ReservationUnitReservationScheduler(
+                reservation_unit, opening_hours_end=end.date()
+            )
+            self.check_opening_hours(scheduler, begin, end)
+            self.check_open_application_round(scheduler, begin, end)
+            self.check_reservation_start_time(scheduler, begin)
+
+        self.check_and_handle_pricing(data)
+
+        if self.instance._requires_handling():
+            data["state"] = STATE_CHOICES.REQUIRES_HANDLING
+
+        return data
+
+    def check_begin(self, begin, end):
+        if begin > end:
+            raise ValidationErrorWithCode(
+                "End cannot be before begin",
+                ValidationErrorCodes.RESERVATION_MODIFICATION_NOT_ALLOWED,
+            )
+
+        now = datetime.datetime.now(tz=DEFAULT_TIMEZONE)
+
+        if begin < now:
+            raise ValidationErrorWithCode(
+                "Reservation new begin cannot be in the past",
+                ValidationErrorCodes.RESERVATION_BEGIN_IN_PAST,
+            )
+
+        if self.instance.begin < now:
+            raise ValidationErrorWithCode(
+                "Reservation time cannot be changed when current begin time is in past.",
+                ValidationErrorCodes.RESERVATION_CURRENT_BEGIN_IN_PAST,
+            )
+
+    def check_cancellation_rules(self, reservation_unit):
+        now = datetime.datetime.now(tz=DEFAULT_TIMEZONE)
+
+        cancel_rule = reservation_unit.cancellation_rule
+        if not cancel_rule:
+            raise ValidationErrorWithCode(
+                "Reservation time cannot be changed thus no cancellation rule.",
+                ValidationErrorCodes.CANCELLATION_NOT_ALLOWED,
```

Comment:
Here, too, we have multiple errors with `CANCELLATION_NOT_ALLOWED` error codes. Can reservation be cancelled via time change mutation? 🤔

Reply:
The error code itself is specsed in ticket to be `CANCELLATION_NOT_ALLOWED` since here is the cancellation rule that is causing the fail. We first pondered that should we return different codes for every situation but nah, the UI knows that it is cancellation rule related error when changing time when return `CANCELLATION_NOT_ALLOWED`

So shortly; the time change itself is kind of a cancellation for the old time.